cpp/include/cuml/experimental/fil/detail/infer_kernel/gpu.cuh

```diff
@@ -167,54 +169,80 @@ infer_kernel(
       auto tree_output = std::conditional_t<
         has_vector_leaves, typename node_t::index_type, typename node_t::threshold_type
       >{};
-      if constexpr (has_nonlocal_categories) {
-        tree_output = evaluate_tree<has_vector_leaves>(
-          forest.get_tree_root(tree_index),
-          input_data + row_index * col_count,
-          categorical_data
-        );
+      auto leaf_node_id = index_type{};
```

Comment:
Same questions as for the CPU section.

Aside: Now that this logic is a little more complex, we should definitely try to unify it across CPU/GPU. The indexing logic is identical, so it would be good to have it in a single place. Definitely not for this PR, though.